File: AiSnapper/OpenAIClient.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace AiSnapper
{
    public static class OpenAIClient
    {
        private static readonly HttpClient _http = new HttpClient();
        private const string ApiUrl = "https://api.openai.com/v1/chat/completions"; // vision via content parts
        private const string DefaultModel = "gpt-4o"; // best general vision model

        public static async Task<string> AskAsync(string prompt, string base64Png)
        {
            var apiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY");
            if (string.IsNullOrEmpty(apiKey))
                throw new InvalidOperationException("OPENAI_API_KEY environment variable is not set.");

            var model = Environment.GetEnvironmentVariable("OPENAI_MODEL") ?? DefaultModel;

            _http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);

            // Build a vision message: text + image_url (data URI). image_url must be an object with url/detail.
            var payload = new
            {
                model = model,
                messages = new object[]
                {
                    new {
                        role = "user",
                        content = new object[] {
                            new { type = "text", text = prompt },
                            new { type = "image_url", image_url = new { url = $"data:image/png;base64,{base64Png}", detail = "high" } }
                        }
                    }
                },
                temperature = 0.2
            };

            var json = JsonSerializer.Serialize(payload);
            var res = await _http.PostAsync(ApiUrl, new StringContent(json, Encoding.UTF8, "application/json"));
            var body = await res.Content.ReadAsStringAsync();
            if (!res.IsSuccessStatusCode)
            {
                throw new InvalidOperationException($"OpenAI error {res.StatusCode}: {body}");
            }

            using var doc = JsonDocument.Parse(body);
            var root = doc.RootElement;
            var text = root.GetProperty("choices")[0].GetProperty("message").GetProperty("content").GetString();
            return text ?? "";
        }
    }
}

File: AiSnapperMac/src/openai_client.py
import os
import json
import base64
import requests
from typing import List, Callable, Optional

API_URL = "https://api.openai.com/v1/chat/completions"
DEFAULT_MODEL = os.getenv("OPENAI_MODEL", "gpt-4o")

class OpenAIClient:
    def __init__(self, api_key: Optional[str] = None, model: Optional[str] = None):
        self.api_key = api_key or os.getenv("OPENAI_API_KEY")
        if not self.api_key:
            raise RuntimeError("OPENAI_API_KEY is not set")
        self.model = model or DEFAULT_MODEL
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        })

    def ask(self, messages: List[dict]) -> str:
        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": 0.2
        }
        r = self.session.post(API_URL, data=json.dumps(payload))
        if r.status_code != 200:
            raise RuntimeError(f"OpenAI error {r.status_code}: {r.text}")
        data = r.json()
        return data["choices"][0]["message"]["content"]

    def ask_stream(self, messages: List[dict], on_delta: Callable[[str], None]):
        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": 0.2,
            "stream": True
        }
        with self.session.post(API_URL, data=json.dumps(payload), stream=True) as r:
            if r.status_code != 200:
                raise RuntimeError(f"OpenAI error {r.status_code}: {r.text}")
            for line in r.iter_lines(decode_unicode=True):
                if not line:
                    continue
                if line.startswith("data: "):
                    data = line[6:].strip()
                    if data == "[DONE]":
                        break
                    try:
                        obj = json.loads(data)
                        delta = obj["choices"][0]["delta"].get("content")
                        if isinstance(delta, str):
                            on_delta(delta)
                    except Exception:
                        pass


def image_message(prompt: str, png_bytes: bytes) -> List[dict]:
    b64 = base64.b64encode(png_bytes).decode("ascii")
    return [
        {"role": "system", "content": [ {"type": "text", "text": "You are a helpful assistant. Be concise and reference the image when relevant."} ]},
        {
            "role": "user",
            "content": [
                {"type": "text", "text": prompt},
                {"type": "image_url", "image_url": {"url": f"data:image/png;base64,{b64}", "detail": "high"}}
            ]
        }
    ]
